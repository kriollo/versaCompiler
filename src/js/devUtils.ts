import { html } from 'code-tag';

/**
 * Sanitize the module path to prevent directory traversal attacks.
 * @param {string} module - The module path to sanitize.
 * @returns {string} - The sanitized module path.
 */
export const sanitizeModulePath = (module: string): string => {
    return module
        .replace(/\.\.\//g, '') // Eliminar ".."
        .replace(/\/+/g, '/') // Normalizar barras
        .replace(/[^a-zA-Z0-9/_-]/g, ''); // Eliminar caracteres no permitidos
};

type ModuleName = string & { __brand: 'ModuleName' };
export function isValidModuleName(module: string): module is ModuleName {
    const MODULE_NAME_REGEX = /^(?:@\/)?[a-zA-Z0-9][a-zA-Z0-9/_-]*[a-zA-Z0-9]$/;
    return MODULE_NAME_REGEX.test(module);
}

/**
 * Manejar errores durante la carga del módulo.
 * @param {any} error - El error ocurrido.
 * @param {string} module - El módulo que causó el error.
 */
export function handleError(
    error: any,
    module: string,
    container: HTMLElement,
): void {
    const errorMessage =
        error instanceof Error ? error.message : 'Error desconocido';
    const moduleInfo = module ? `Módulo: ${module}` : 'Módulo no especificado';
    const safeMessage = `${moduleInfo}<br>${errorMessage}`
        .replace(/</g, '<')
        .replace(/>/g, '>');

    // Mostrar mensaje de error en el contenedor
    container.textContent = ''; // Limpiar contenido previo
    container.insertAdjacentHTML(
        'beforeend',
        html`
            <div style="padding: 1rem; background-color: #f44336; color: #fff;">
                <h2>Error crítico</h2>
                <p style="text-align: left;">${safeMessage}</p>
                <button
                    style="padding: 0.5rem 1rem; background-color: #fff; color: #333; border: 1px solid #333; cursor: pointer;"
                    onclick="window.location.reload()">
                    Reintentar
                </button>
            </div>
        `,
    );

    console.error('[Module Loader]', error);
}

export const $dom = (
    selector: string,
    context: Document | Element = document,
): Element | null => context.querySelector(selector);
