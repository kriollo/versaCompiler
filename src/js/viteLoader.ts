import { createApp, defineAsyncComponent } from 'vue';
// Add Vite client types for import.meta.glob
/// <reference types="vite/client" />

// Detectar el m√≥dulo desde la URL (?modulo=otroModulo/dashModule)
const url = new URL(import.meta.url);
const urlParams = url.search;
const searchParams = new URLSearchParams(urlParams);
const modulo = searchParams.get('modulo') || 'appLoader';

// Importar todos los .vue recursivamente
const components = import.meta.glob('@/js/module/**/*.vue');

// Resolver la clave correcta
const path = `/src/js/module/${modulo}.vue`;

if (!(path in components)) {
    throw new Error(`Componente no encontrado: ${modulo}`);
}

const AsyncComponent = defineAsyncComponent(components[path]);

const app = createApp({
    components: { AsyncComponent },
    template: `<AsyncComponent />`,
});

app.mount('#app');
