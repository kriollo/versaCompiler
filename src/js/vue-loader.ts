import {
    handleError,
    isValidModuleName,
    sanitizeModulePath,
    $dom,
} from '@/js/devUtils';
import { createApp, ref } from 'vue';

const debug = ref(false);
const $contenedor = ref(null);

async function loadModule() {
    try {
        const url = new URL(import.meta.url);
        const urlParams = url.search;
        const searchParams = new URLSearchParams(urlParams);
        let module = searchParams.get('m');

        $contenedor.value = $dom('#app');

        if (!$contenedor.value) {
            throw new Error(
                'No se ha encontrado el contenedor para cargar el módulo.',
            );
        }

        if (!module) {
            throw new Error('No se ha especificado un módulo para cargar.');
        }
        module = sanitizeModulePath(module);
        module = module.startsWith('/') ? module.slice(1) : module;
        // Validar el parámetro del módulo
        if (!isValidModuleName(module)) {
            throw new Error(
                'El parámetro del módulo contiene caracteres no permitidos.',
            );
        }

        const component = module.split('/').pop();
        // Importar dinámicamente el módulo
        const moduleResponse = await import(`@/${module}.js?v=${Date.now()}`);
        if (moduleResponse) {
            // Montar el módulo en el contenedor
            const app = createApp({
                components: { [component]: moduleResponse.default },
                setup() {
                    // Configuración de la aplicación según el modo de depuración
                    if (debug.value) {
                        console.log('Debug mode is enabled');
                    }
                    const componentKey = ref(Date.now());
                    return {
                        componentKey,
                    };
                },
                name: 'App',
                template: `<${component} :key="componentKey" />`,
            });

            // Configuración de la aplicación según el modo de depuración
            if (debug) {
                app.config.warnHandler = function (msg, vm, trace) {
                    console.warn(msg, vm, trace);
                };
                app.config.errorHandler = function (err, vm, info) {
                    console.error(err, vm, info);
                };
                app.config.compilerOptions.comments = true;
            } else {
                app.config.compilerOptions.comments = false;
            }
            app.config.performance = true;
            app.config.compilerOptions.whitespace = 'condense';

            app.mount($contenedor.value, true);

            // --- FRAGMENTO CLAVE PARA HMR Y ACCESO GLOBAL ---
            if (typeof window !== 'undefined') {
                // Exponemos la instancia de la app para HMR y depuración
                (window as any).__VUE_APP__ = app;
                // Exponemos el proxy raíz, útil para acceder a métodos, $forceUpdate, etc.
                (window as any).__VUE_APP_PROXY__ = app._instance?.proxy;
            }
        }
    } catch (e) {
        handleError(e, module, $contenedor);
    }
}

loadModule();
