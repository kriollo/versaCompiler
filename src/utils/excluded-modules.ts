/**
 * Lista centralizada de módulos excluidos de la resolución automática
 * Estos módulos se mantienen con su importación original sin transformar
 */

export const EXCLUDED_MODULES = new Set([
    // Módulos de Vue.js que requieren resolución específica
    'vue/compiler-sfc',
    'vue/dist/vue.runtime.esm-bundler',
    '@vue/compiler-sfc',
    '@vue/compiler-dom',
    '@vue/runtime-core',
    '@vue/runtime-dom',

    // Módulos de oxc-parser que tienen dependencias específicas de WASM
    'oxc-parser',
    'oxc-parser/wasm',
    'oxc-minify',
    'oxc-minify/browser',
    '@oxc-parser/binding-wasm32-wasi',
    '@oxc-minify/binding-wasm32-wasi',

    // Módulos de TypeScript que pueden tener resoluciones complejas
    'typescript',
    'typescript/lib/typescript',

    // Módulos de herramientas de build y utilidades que deben mantenerse originales
    'yargs',
    'yargs/helpers',
    'yargs-parser',
    'chalk',
    'browser-sync',
    'chokidar',
    'get-port',
    'execa',
    'find-root',
    'fs-extra',
    'minimatch', // ✅ Incluir minimatch aquí
]);
