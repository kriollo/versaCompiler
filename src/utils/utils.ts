/**
 * Converts a 24-hour time string to a 12-hour time string with AM/PM.
 *
 * @param {number} timing - The value of the timing en miliseconds.
 * @returns {string} the timing in ms, seconds, minutes or hours.
 */
export const showTimingForHumans = (timing: any): string => {
    if (timing < 1000) {
        return `${timing} ms`;
    } else if (timing < 60000) {
        return `${timing / 1000} s`;
    } else if (timing < 3600000) {
        return `${timing / 60000} min`;
    } else {
        return `${timing / 3600000} h`;
    }
};
