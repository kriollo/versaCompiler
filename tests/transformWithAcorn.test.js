import { transformModuleWithAcorn } from '../dist/utils/transformWithAcorn';

describe('transformModuleWithAcorn', () => {
    test('debería transformar un import .js local a dinámico y añadir HMR', () => {
        const inputCode = `
import { myUtil } from './utils.js';

myUtil();
console.log('Hola desde el módulo');
        `;
        expect(transformModuleWithAcorn(inputCode)).toMatchSnapshot();
    });

    test('NO debería transformar imports si el archivo contiene "defineComponent" y debería añadir marca de recarga', () => {
        const inputCode = `
import { ref } from 'vue';
import { anotherUtil } from './another.js';
import { defineComponent } from 'vue';

export default defineComponent({
  setup() {
    const count = ref(0);
    anotherUtil();
    return { count };
  }
});
        `;
        expect(transformModuleWithAcorn(inputCode)).toMatchSnapshot();
    });

    test('NO debería transformar imports si el archivo contiene "createApp" y debería añadir marca de recarga', () => {
        const inputCode = `
import { createApp } from 'vue';
import App from './App.vue';
import { router } from './router.js';

const app = createApp(App);
app.use(router);
app.mount('#app');
        `;
        expect(transformModuleWithAcorn(inputCode)).toMatchSnapshot();
    });

    test('NO debería transformar imports si el archivo contiene "mount" (y no createApp) y debería añadir marca de recarga', () => {
        const inputCode = `
import { h, render } from 'vue'; // Asumiendo que no usa createApp directamente
import App from './App.vue';

// Lógica de montaje más manual
const vnode = h(App);
render(vnode, document.getElementById('#app'));
// Simulación de una condición que podría llevar a la marca de recarga
// si 'mount' fuera el único indicador.
// En la lógica actual, esto no activaría isVueInitializationFile por sí solo
// pero si 'mount' se añade explícitamente a la condición, sí lo haría.
// Para este test, asumimos que la lógica de 'isVueInitializationFile'
// considera '.mount('#app')' o similar.
// Si la lógica es estrictamente code.includes('mount'), entonces:
const someObject = { mount: () => {} };
someObject.mount();
        `;
        // Este snapshot reflejará cómo la lógica actual de isVueInitializationFile
        // (que busca code.includes('mount')) trata este caso.
        expect(transformModuleWithAcorn(inputCode)).toMatchSnapshot();
    });

    test('NO debería transformar imports de librerías externas como "vue"', () => {
        const inputCode = `
import { ref, computed } from 'vue';
import { localUtil } from './local.js';

const count = ref(0);
const double = computed(() => count.value * 2);
localUtil();
        `;
        // Esperamos que local.js se transforme, pero vue no.
        expect(transformModuleWithAcorn(inputCode)).toMatchSnapshot();
    });

    test('NO debería transformar un archivo que es principalmente imports/exports (CoreDefinitionFile) y debería añadir marca de recarga', () => {
        const inputCode = `
import { firstUtil } from './utils/first.js';
import { secondUtil } from './utils/second.js';
import { thirdUtil } from './utils/third.js'; // .js local
// Alguna lógica mínima o ninguna
const combined = () => firstUtil() + secondUtil();

export { firstUtil, secondUtil, thirdUtil, combined };
export const CONST_A = 123;
        `;
        // Asumiendo que la heurística de ratio (e.g., >=0.7) lo marca como CoreDefinitionFile
        // y que './utils/third.js' cuenta como hasLocalJsImport.
        expect(transformModuleWithAcorn(inputCode)).toMatchSnapshot();
    });

    test('debería manejar un archivo solo con exports', () => {
        const inputCode = `
export const message = "Hola";
export function greet(name) {
  return \`Hola, \${name}\`;
}
        `;
        // No hay imports dinámicos, no debería haber IIFE ni HMR.
        // No debería tener la marca de recarga a menos que cumpla otra condición.
        expect(transformModuleWithAcorn(inputCode)).toMatchSnapshot();
    });

    test('debería manejar un archivo vacío o solo con comentarios', () => {
        const inputCode = `
// Esto es un comentario
// Otro comentario
        `;
        expect(transformModuleWithAcorn(inputCode)).toMatchSnapshot();

        const emptyCode = ``;
        expect(transformModuleWithAcorn(emptyCode)).toMatchSnapshot();
    });

    test('debería mover el cuerpo del código dentro del IIFE si hay imports dinámicos', () => {
        const inputCode = `
import { $dom } from './dom-utils.js';

const button = $dom('#myButton');
button.addEventListener('click', () => {
    console.log('Button clicked!');
});
export const testVar = 123;
        `;
        expect(transformModuleWithAcorn(inputCode)).toMatchSnapshot();
    });

    test('debería transformar diferentes tipos de imports .js locales', () => {
        const inputCode = `
import myDefault from './defaultUtil.js';
import * as allUtils from './allUtils.js';
import { specificUtil, originalName as aliasUtil } from './specificUtils.js';

myDefault();
allUtils.doSomething();
specificUtil();
aliasUtil();
        `;
        expect(transformModuleWithAcorn(inputCode)).toMatchSnapshot();
    });

    test('NO debería transformar un import .js si su variable se usa en _resolveComponent, pero SÍ otros imports .js', () => {
        const inputCode = `
import MyComponent from './MyComponent.js'; // Usado en _resolveComponent
import anotherUtil from './anotherUtil.js'; // No usado en _resolveComponent
import { defineComponent } from 'vue';

export default defineComponent({
  components: {
    'my-component': _resolveComponent("MyComponent")
  },
  setup() {
    anotherUtil();
    return {};
  }
});
        `;
        // Esperamos que MyComponent.js NO se transforme (se mantenga estático)
        // pero anotherUtil.js SÍ se transforme a dinámico.
        // El archivo general NO debería tener la marca //versaHRM-reloadFILE porque contiene defineComponent,
        // pero la lógica de _resolveComponent es una capa adicional de no-transformación para imports específicos.
        // Sin embargo, la presencia de defineComponent hará que TODO el archivo sea marcado para recarga.
        // Este test verifica principalmente que _resolveComponent no cause errores y que la marca se aplique correctamente.
        expect(transformModuleWithAcorn(inputCode)).toMatchSnapshot();
    });

    test('NO debería transformar un archivo que está justo por debajo del ratio de CoreDefinitionFile', () => {
        const inputCode = `
import { utilOne } from './utils/utilOne.js'; // .js local
import { utilTwo } from './utils/utilTwo.js'; // .js local
export { utilOne, utilTwo };

function localLogic1() { /* ... */ }
function localLogic2() { /* ... */ }
const localValue = 123;

// 2 imports, 1 export (3 nodos import/export)
// 3 nodos de lógica local
// Total nodos = 6
// Ratio = 3/6 = 0.5. Si el umbral es 0.7, esto NO es CoreDefinitionFile.
// Por lo tanto, utilOne.js y utilTwo.js deberían transformarse.
        `;
        expect(transformModuleWithAcorn(inputCode)).toMatchSnapshot();
    });

    test('debería manejar un archivo solo con imports .js locales y sin exports', () => {
        const inputCode = `
import './sideEffectModule1.js';
import { setup } from './setupModule.js';

setup();
console.log('Módulo de solo imports ejecutado');
        `;
        // Esperamos que ambos imports se transformen a dinámicos.
        expect(transformModuleWithAcorn(inputCode)).toMatchSnapshot();
    });

    test('debería preservar comentarios en el código', () => {
        const inputCode = `
// Comentario al inicio del archivo
import { myUtil } from './utils.js'; // Comentario junto al import

/**
 * Comentario de bloque
 */
function doWork() {
    // Comentario dentro de una función
    myUtil();
}

// Comentario al final
export { doWork };
        `;
        // Acorn debería preservar los comentarios. El snapshot lo verificará.
        expect(transformModuleWithAcorn(inputCode)).toMatchSnapshot();
    });

    test('debería manejar re-exportaciones de módulos .js locales (considerándolos CoreDefinitionFile)', () => {
        const inputCode = `
export { default as myDefault } from './myDefaultExport.js';
export * from './myNamespaceExports.js';
export { namedExport, originalName as localReExport } from './myNamedExports.js';
// Este archivo es puramente de re-exportación de módulos locales .js
// Debería ser marcado como CoreDefinitionFile y no transformado, añadiendo la marca de recarga.
        `;
        expect(transformModuleWithAcorn(inputCode)).toMatchSnapshot();
    });

    test('debería manejar re-exportaciones de librerías externas (manteniendo imports estáticos)', () => {
        const inputCode = `
export { ref, computed } from 'vue';
export { default as anotherDefault } from './anotherDefault.js'; // .js local
// El import de 'vue' debe permanecer estático.
// El import de './anotherDefault.js' debería transformarse.
// El archivo en sí no debería ser CoreDefinitionFile solo por re-exportar 'vue'.
        `;
        expect(transformModuleWithAcorn(inputCode)).toMatchSnapshot();
    });


});
