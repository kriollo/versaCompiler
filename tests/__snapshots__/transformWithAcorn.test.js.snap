// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`transformModuleWithAcorn NO debería transformar imports de librerías externas como "vue" 1`] = `
"let ref;
let computed;
let localUtil;

(async () => {
    const importWithTimestamp = path => { const url = \`\${path}?t=\${Date.now()}\`; return import(url); };
    try {
        ({ ref, computed } = await importWithTimestamp('vue'));
        ({ localUtil } = await importWithTimestamp('./local.js'));
        console.log('[HMR] Módulos dinámicos cargados inicialmente');
    } catch (e) {
        console.error('[HMR] Error en carga inicial de módulos dinámicos', e);
    }
    window.__VERSA_HMR = window.__VERSA_HMR || {};
    window.__VERSA_HMR.modules = window.__VERSA_HMR.modules || {};
    window.__VERSA_HMR.modules['vue'] = async () => {
            try {
                const mod = await importWithTimestamp('vue');
                ref = mod.ref;
                computed = mod.computed;
                console.log('[HMR] Módulo vue (named) recargado con éxito');
                return true;
            } catch (e) {
                console.error('[HMR] Error recargando vue', e);
                return false;
            }
        };
    window.__VERSA_HMR.modules['./local.js'] = async () => {
            try {
                const mod = await importWithTimestamp('./local.js');
                localUtil = mod.localUtil;
                console.log('[HMR] Módulo ./local.js (named) recargado con éxito');
                return true;
            } catch (e) {
                console.error('[HMR] Error recargando ./local.js', e);
                return false;
            }
        };
    window.__VERSA_HMR.reload = async () => { console.log("[HMR] Reload all invoked"); const P = Object.values(window.__VERSA_HMR.modules).map(f=>f()); await Promise.all(P); };
const count = ref(0);
    const double = computed(() => count.value * 2);
    localUtil();
})();"
`;

exports[`transformModuleWithAcorn NO debería transformar imports si el archivo contiene "createApp" y debería añadir marca de recarga 1`] = `
"//versaHRM-reloadFILE
import { createApp } from 'vue';
import App from './App.vue';
import { router } from './router.js';

const app = createApp(App);
app.use(router);
app.mount('#app');"
`;

exports[`transformModuleWithAcorn NO debería transformar imports si el archivo contiene "mount" (y no createApp) y debería añadir marca de recarga 1`] = `
"//versaHRM-reloadFILE
import { h, render } from 'vue';
import App from './App.vue';

const vnode = h(App);
render(vnode, document.getElementById('#app'));
const someObject = { mount: () => {} };
someObject.mount();"
`;

exports[`transformModuleWithAcorn NO debería transformar un archivo que es principalmente imports/exports (CoreDefinitionFile) y debería añadir marca de recarga 1`] = `
"import { firstUtil } from './utils/first.js';
import { secondUtil } from './utils/second.js';
import { thirdUtil } from './utils/third.js';

(async () => {
const combined = () => firstUtil() + secondUtil();
    export { firstUtil, secondUtil, thirdUtil, combined };
    export const CONST_A = 123;
})();"
`;

exports[`transformModuleWithAcorn NO debería transformar un archivo que está justo por debajo del ratio de CoreDefinitionFile 1`] = `
"let utilOne;
let utilTwo;

(async () => {
    const importWithTimestamp = path => { const url = \`\${path}?t=\${Date.now()}\`; return import(url); };
    try {
        ({ utilOne } = await importWithTimestamp('./utils/utilOne.js'));
        ({ utilTwo } = await importWithTimestamp('./utils/utilTwo.js'));
        console.log('[HMR] Módulos dinámicos cargados inicialmente');
    } catch (e) {
        console.error('[HMR] Error en carga inicial de módulos dinámicos', e);
    }
    window.__VERSA_HMR = window.__VERSA_HMR || {};
    window.__VERSA_HMR.modules = window.__VERSA_HMR.modules || {};
    window.__VERSA_HMR.modules['./utils/utilOne.js'] = async () => {
            try {
                const mod = await importWithTimestamp('./utils/utilOne.js');
                utilOne = mod.utilOne;
                console.log('[HMR] Módulo ./utils/utilOne.js (named) recargado con éxito');
                return true;
            } catch (e) {
                console.error('[HMR] Error recargando ./utils/utilOne.js', e);
                return false;
            }
        };
    window.__VERSA_HMR.modules['./utils/utilTwo.js'] = async () => {
            try {
                const mod = await importWithTimestamp('./utils/utilTwo.js');
                utilTwo = mod.utilTwo;
                console.log('[HMR] Módulo ./utils/utilTwo.js (named) recargado con éxito');
                return true;
            } catch (e) {
                console.error('[HMR] Error recargando ./utils/utilTwo.js', e);
                return false;
            }
        };
    window.__VERSA_HMR.reload = async () => { console.log("[HMR] Reload all invoked"); const P = Object.values(window.__VERSA_HMR.modules).map(f=>f()); await Promise.all(P); };
export { utilOne, utilTwo };
    function localLogic1() { /* ... */ }
    function localLogic2() { /* ... */ }
    const localValue = 123;
})();"
`;

exports[`transformModuleWithAcorn componente Vue con _resolveComponent: import resuelto no se transforma, otro .js local sí, sin marca de recarga 1`] = `
"import MyComponentImported from './MyComponentImported.js';

export default defineComponent({
  components: {
    'my-component': _resolveComponent("MyComponentImported")
  },
  setup() {
    anotherUtil(); // Se usa la utilidad local
    return {};
  }
});    let anotherUtil;
    let defineComponent;
    let _resolveComponent;
    const importWithTimestamp = path => { const url = \`\${path}?t=\${Date.now()}\`; return import(url); };
    const load___anotherUtil_js_module = async () => {
    const mod = await importWithTimestamp('./anotherUtil.js');
    anotherUtil = mod.default;
    };
    const load_vue_module = async () => {
    const mod = await importWithTimestamp('vue');
    defineComponent = mod.defineComponent;
    _resolveComponent = mod._resolveComponent;
    };
    await load___anotherUtil_js_module();
    await load_vue_module();
    window.__VERSA_HMR = window.__VERSA_HMR || {};
    window.__VERSA_HMR.modules = window.__VERSA_HMR.modules || {}
    window.__VERSA_HMR.modules['./anotherUtil.js'] = async () => {
    try {
    await load___anotherUtil_js_module();
    console.log('[HMR] Módulo recargado: ./anotherUtil.js');
    return true;
    } catch (e) {
    console.error('[HMR] Falló recarga ./anotherUtil.js', e);
    return false;
    }
    };
    window.__VERSA_HMR.modules['vue'] = async () => {
    try {
    await load_vue_module();
    console.log('[HMR] Módulo recargado: vue');
    return true;
    } catch (e) {
    console.error('[HMR] Falló recarga vue', e);
    return false;
    }
    };"
`;

exports[`transformModuleWithAcorn componente Vue: imports .js locales se transforman, cuerpo fuera de IIFE, sin marca de recarga 1`] = `
"export default defineComponent({
  name: 'MiComponente',
  components: { UnComponenteHijo },
  setup() {
    const mensaje = ref("Hola");
    utilidadLocal(); // Se usa la utilidad local
    const renderHijo = () => h(UnComponenteHijo);
    return { mensaje, renderHijo };
  }
});    let ref;
    let defineComponent;
    let h;
    let utilidadLocal;
    let UnComponenteHijo;
    const importWithTimestamp = path => { const url = \`\${path}?t=\${Date.now()}\`; return import(url); };
    const load___UnComponenteHijo_vue_module = async () => {
    const mod = await importWithTimestamp('./UnComponenteHijo.vue');
    UnComponenteHijo = mod.default;
    };
    const load_vue_module = async () => {
    const mod = await importWithTimestamp('vue');
    ref = mod.ref;
    defineComponent = mod.defineComponent;
    h = mod.h;
    };
    const load___utilidadLocal_js_module = async () => {
    const mod = await importWithTimestamp('./utilidadLocal.js');
    utilidadLocal = mod.utilidadLocal;
    };
    await load___UnComponenteHijo_vue_module();
    await load_vue_module();
    await load___utilidadLocal_js_module();
    window.__VERSA_HMR = window.__VERSA_HMR || {};
    window.__VERSA_HMR.modules = window.__VERSA_HMR.modules || {}
    window.__VERSA_HMR.modules['./UnComponenteHijo.vue'] = async () => {
    try {
    await load___UnComponenteHijo_vue_module();
    console.log('[HMR] Módulo recargado: ./UnComponenteHijo.vue');
    return true;
    } catch (e) {
    console.error('[HMR] Falló recarga ./UnComponenteHijo.vue', e);
    return false;
    }
    };
    window.__VERSA_HMR.modules['vue'] = async () => {
    try {
    await load_vue_module();
    console.log('[HMR] Módulo recargado: vue');
    return true;
    } catch (e) {
    console.error('[HMR] Falló recarga vue', e);
    return false;
    }
    };
    window.__VERSA_HMR.modules['./utilidadLocal.js'] = async () => {
    try {
    await load___utilidadLocal_js_module();
    console.log('[HMR] Módulo recargado: ./utilidadLocal.js');
    return true;
    } catch (e) {
    console.error('[HMR] Falló recarga ./utilidadLocal.js', e);
    return false;
    }
    };"
`;

exports[`transformModuleWithAcorn debería manejar re-exportaciones de librerías externas (manteniendo imports estáticos) 1`] = `
"(async () => {
export { ref, computed } from 'vue';
    export { default as anotherDefault } from './anotherDefault.js';
})();"
`;

exports[`transformModuleWithAcorn debería manejar re-exportaciones de módulos .js locales (considerándolos CoreDefinitionFile) 1`] = `
"(async () => {
export { default as myDefault } from './myDefaultExport.js';
    export * from './myNamespaceExports.js';
    export { namedExport, originalName as localReExport } from './myNamedExports.js';
})();"
`;

exports[`transformModuleWithAcorn debería manejar un archivo solo con exports 1`] = `
"(async () => {
export const message = "Hola";
    export function greet(name) {
      return \`Hola, \${name}\`;
    }
})();"
`;

exports[`transformModuleWithAcorn debería manejar un archivo solo con imports .js locales y sin exports 1`] = `
"let setup;

(async () => {
    const importWithTimestamp = path => { const url = \`\${path}?t=\${Date.now()}\`; return import(url); };
    try {
        ({ setup } = await importWithTimestamp('./setupModule.js'));
        console.log('[HMR] Módulos dinámicos cargados inicialmente');
    } catch (e) {
        console.error('[HMR] Error en carga inicial de módulos dinámicos', e);
    }
    window.__VERSA_HMR = window.__VERSA_HMR || {};
    window.__VERSA_HMR.modules = window.__VERSA_HMR.modules || {};
    window.__VERSA_HMR.modules['./setupModule.js'] = async () => {
            try {
                const mod = await importWithTimestamp('./setupModule.js');
                setup = mod.setup;
                console.log('[HMR] Módulo ./setupModule.js (named) recargado con éxito');
                return true;
            } catch (e) {
                console.error('[HMR] Error recargando ./setupModule.js', e);
                return false;
            }
        };
    window.__VERSA_HMR.reload = async () => { console.log("[HMR] Reload all invoked"); const P = Object.values(window.__VERSA_HMR.modules).map(f=>f()); await Promise.all(P); };
setup();
    console.log('Módulo de solo imports ejecutado');
})();"
`;

exports[`transformModuleWithAcorn debería manejar un archivo vacío o solo con comentarios 1`] = `""`;

exports[`transformModuleWithAcorn debería manejar un archivo vacío o solo con comentarios 2`] = `""`;

exports[`transformModuleWithAcorn debería mover el cuerpo del código dentro del IIFE si hay imports dinámicos 1`] = `
"let $dom;

(async () => {
    const importWithTimestamp = path => { const url = \`\${path}?t=\${Date.now()}\`; return import(url); };
    try {
        ({ $dom } = await importWithTimestamp('./dom-utils.js'));
        console.log('[HMR] Módulos dinámicos cargados inicialmente');
    } catch (e) {
        console.error('[HMR] Error en carga inicial de módulos dinámicos', e);
    }
    window.__VERSA_HMR = window.__VERSA_HMR || {};
    window.__VERSA_HMR.modules = window.__VERSA_HMR.modules || {};
    window.__VERSA_HMR.modules['./dom-utils.js'] = async () => {
            try {
                const mod = await importWithTimestamp('./dom-utils.js');
                $dom = mod.$dom;
                console.log('[HMR] Módulo ./dom-utils.js (named) recargado con éxito');
                return true;
            } catch (e) {
                console.error('[HMR] Error recargando ./dom-utils.js', e);
                return false;
            }
        };
    window.__VERSA_HMR.reload = async () => { console.log("[HMR] Reload all invoked"); const P = Object.values(window.__VERSA_HMR.modules).map(f=>f()); await Promise.all(P); };
const button = $dom('#myButton');
    button.addEventListener('click', () => {
        console.log('Button clicked!');
    });
    export const testVar = 123;
})();"
`;

exports[`transformModuleWithAcorn debería preservar comentarios en el código 1`] = `
"let myUtil;

(async () => {
    const importWithTimestamp = path => { const url = \`\${path}?t=\${Date.now()}\`; return import(url); };
    try {
        ({ myUtil } = await importWithTimestamp('./utils.js'));
        console.log('[HMR] Módulos dinámicos cargados inicialmente');
    } catch (e) {
        console.error('[HMR] Error en carga inicial de módulos dinámicos', e);
    }
    window.__VERSA_HMR = window.__VERSA_HMR || {};
    window.__VERSA_HMR.modules = window.__VERSA_HMR.modules || {};
    window.__VERSA_HMR.modules['./utils.js'] = async () => {
            try {
                const mod = await importWithTimestamp('./utils.js');
                myUtil = mod.myUtil;
                console.log('[HMR] Módulo ./utils.js (named) recargado con éxito');
                return true;
            } catch (e) {
                console.error('[HMR] Error recargando ./utils.js', e);
                return false;
            }
        };
    window.__VERSA_HMR.reload = async () => { console.log("[HMR] Reload all invoked"); const P = Object.values(window.__VERSA_HMR.modules).map(f=>f()); await Promise.all(P); };
function doWork() {
        // Comentario dentro de una función
        myUtil();
    }
    export { doWork };
})();"
`;

exports[`transformModuleWithAcorn debería transformar diferentes tipos de imports .js locales 1`] = `
"let myDefault;
let allUtils;
let specificUtil;
let aliasUtil;

(async () => {
    const importWithTimestamp = path => { const url = \`\${path}?t=\${Date.now()}\`; return import(url); };
    try {
        myDefault = (await importWithTimestamp('./defaultUtil.js')).default;
        allUtils = await importWithTimestamp('./allUtils.js');
        ({ specificUtil, aliasUtil } = await importWithTimestamp('./specificUtils.js'));
        console.log('[HMR] Módulos dinámicos cargados inicialmente');
    } catch (e) {
        console.error('[HMR] Error en carga inicial de módulos dinámicos', e);
    }
    window.__VERSA_HMR = window.__VERSA_HMR || {};
    window.__VERSA_HMR.modules = window.__VERSA_HMR.modules || {};
    window.__VERSA_HMR.modules['./defaultUtil.js'] = async () => {
            try {
                const mod = await importWithTimestamp('./defaultUtil.js');
                myDefault = mod.default;
                console.log('[HMR] Módulo ./defaultUtil.js (default) recargado con éxito');
                return true;
            } catch (e) {
                console.error('[HMR] Error recargando ./defaultUtil.js', e);
                return false;
            }
        };
    window.__VERSA_HMR.modules['./allUtils.js'] = async () => {
            try {
                const mod = await importWithTimestamp('./allUtils.js');
                allUtils = mod;
                console.log('[HMR] Módulo ./allUtils.js (namespace) recargado con éxito');
                return true;
            } catch (e) {
                console.error('[HMR] Error recargando ./allUtils.js', e);
                return false;
            }
        };
    window.__VERSA_HMR.modules['./specificUtils.js'] = async () => {
            try {
                const mod = await importWithTimestamp('./specificUtils.js');
                specificUtil = mod.specificUtil;
                aliasUtil = mod.originalName;
                console.log('[HMR] Módulo ./specificUtils.js (named) recargado con éxito');
                return true;
            } catch (e) {
                console.error('[HMR] Error recargando ./specificUtils.js', e);
                return false;
            }
        };
    window.__VERSA_HMR.reload = async () => { console.log("[HMR] Reload all invoked"); const P = Object.values(window.__VERSA_HMR.modules).map(f=>f()); await Promise.all(P); };
myDefault();
    allUtils.doSomething();
    specificUtil();
    aliasUtil();
})();"
`;

exports[`transformModuleWithAcorn debería transformar un import .js local a dinámico y añadir HMR 1`] = `
"let myUtil;

(async () => {
    const importWithTimestamp = path => { const url = \`\${path}?t=\${Date.now()}\`; return import(url); };
    try {
        ({ myUtil } = await importWithTimestamp('./utils.js'));
        console.log('[HMR] Módulos dinámicos cargados inicialmente');
    } catch (e) {
        console.error('[HMR] Error en carga inicial de módulos dinámicos', e);
    }
    window.__VERSA_HMR = window.__VERSA_HMR || {};
    window.__VERSA_HMR.modules = window.__VERSA_HMR.modules || {};
    window.__VERSA_HMR.modules['./utils.js'] = async () => {
            try {
                const mod = await importWithTimestamp('./utils.js');
                myUtil = mod.myUtil;
                console.log('[HMR] Módulo ./utils.js (named) recargado con éxito');
                return true;
            } catch (e) {
                console.error('[HMR] Error recargando ./utils.js', e);
                return false;
            }
        };
    window.__VERSA_HMR.reload = async () => { console.log("[HMR] Reload all invoked"); const P = Object.values(window.__VERSA_HMR.modules).map(f=>f()); await Promise.all(P); };
myUtil();
    console.log('Hola desde el módulo');
})();"
`;
